### PR TITLE
Integration tests for `testnet.polykey.io`

### LINKED ISSUE #388: Add diagrams to NAT Traversal tests
### Specification
The NAT traversal tests are complex and rely on many utility functions, making it difficult to know what is happening in these tests. It would be helpful to have diagrams at the top of each test showing the architecture of the namespaces and agents involved in each test.

Diagrams needed:
- DMZ
  - [x] Agent 1 (DMZ) -> Agent 2 (DMZ)
  - [x] Agent 1 (DMZ) -> Seed Node -> Agent 2 (DMZ)
- Endpoint-independent mapping (EIM)
  - [x] Agent 1 (EIM) -> Agent 2 (DMZ)
  - [x] Agent 1 (DMZ) -> Agent 2 (EIM)
  - [x] Agent 1 (EIM) -> Agent 2 (EIM)
  - [x] Agent 1 (EIM) -> Seed Node -> Agent 2 (EIM)
  - [x] Agent 1 (EIM) -> Seed Node -> Agent 2 (EDM)
- Endpoint-dependent mapping (EDM)
  - [x] Agent 1 (EDM) -> Agent 2 (DMZ)
  - [x] Agent 1 (DMZ) -> Agent 2 (EDM)
  - [x] Agent 1 (EDM) -> Seed Node -> Agent 2 (EDM)
  - [x] Agent 1 (EDM) -> Seed Node -> Agent 2 (EIM)

### Additional context
- First suggested here: https://github.com/MatrixAI/js-polykey/pull/381#issuecomment-1154639269

### Tasks
1. Create diagrams for each of the NAT tests using ASCII flow
2. Modify test descriptions where necessary to make them clearer

### LINKED ISSUE #416: Tests for `testnet.polykey.io`
### Specification
We need a suite of tests to cover interacting with a deployed agent, which we can do using `testnet.polykey.io`. These tests need to cover various different connection scenarios:

- Connecting to a deployed agent as a seed node during startup
- Pinging a deployed agent
- Using the deployed agent as a signaller (and eventually a relay once this is implemented)
- Tests for when the deployed agent is contacting an agent behind a NAT (https://github.com/MatrixAI/js-polykey/issues/159)
- Any bugs that are discovered during the above tasks

These tests should go into their own subdirectory `tests/testnet` and should not be run with the other tests. They should be disabled in our jest config and should only run when explicitly called (which will happen during the integration stage of our pipelines).

### Additional context
- https://github.com/MatrixAI/js-polykey/issues/159 - NAT traversal tests against `testnet.polykey.io`
- https://github.com/MatrixAI/js-polykey/issues/414 - Issue for failed connections to/from a deployed agent
- https://github.com/MatrixAI/js-polykey/issues/415 - Issue for continuous re-connection bug when connecting to deployed seed node
- https://github.com/MatrixAI/js-polykey/issues/413 - Issue for potential infinite loop when connecting to deployed seed node

### Tasks
1. Attempt connections to the deployed seed node and create issues for all bugs discovered (and resolve them)
2. Create tests for simple connections to `testnet.polykey.io`
3. Create tests for edge cases and previous bugs
4. Create tests for connecting to a deployed seed node from behind a NAT

### PR DESCRIPTION
**Pull request by emmacasolin**
### Description
Adding a suite of integration tests for interacting with a deployed agent at `testnet.polykey.io`. These will go under `tests/integration/testnet`.

### Issues Fixed
* Related MatrixAI/Polykey-CLI#71 
* Fixes MatrixAI/Polykey#388 

### Tasks
- [x] 1. Add diagrams to existing NAT tests
- [x] 2. Create new subdirectory `tests/integration/testnet`
- ~[ ] 3. Write tests as specified in MatrixAI/Polykey-CLI#71~ 

Required tests:
- [X] Connecting to `testnet.polykey.io` seed node (by setting as seed node during `pk agent start`)
- ~[ ] Verification of seed node in local keynode's `NodeGraph` (`pk nodes getall`)~
- ~[ ] Sending notification to seed node (expect failure because no trust established, but this shouldn't show up locally)~
- ~[ ] Establishing trust of seed node with local keynode (`pk identities trust`)~
- ~[ ] Verifying node has been added to seed node's gestalt graph after trust has been established (`pk identities get/list`)~
- ~[ ] Sending notification to seed node, and verifying seed node received (will need to connect to testnet and run `pk notifications read`)~
- [X] Pinging the seed node
- [X] Pinging another node through the seed node

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

**Comment by CMCDragonkai** [comment]
Ok so these will be the first directory grouped integration tests with respect to MatrixAI/Polykey#435. So far you don't need to "reuse" any tests from `tests/nat` right as MatrixAI/Polykey#435 isn't done yet.

**Comment by emmacasolin** [comment]
How should the testnet be specified as a seed node in these tests? Using `--network=testnet` isn't working for me, and specifying the deployed nodes individually using `--seed-nodes` would be too brittle since the deployed agents change relatively frequently.

**Comment by emmacasolin** [comment]
One problem with our testnet NAT tests is that the testnet needs to coordinate signalling between two network namespaces running on a single real machine. This essentially becomes a double NAT.

```
                                           Testnet
                                      ┌───────────────┐
                                      │               │
                                      │ ┌───────────┐ │
                                      │ │ SeedNode  │ │
                                      │ └───────────┘ │
                                      │   HOST:PORT   │
                                      │               │
                                      └───────────────┘

    Real Computer
┌──────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                          │
│                                       ┌──────────┐                                       │
│                                       │ Real NAT │                                       │
│                                       └──────────┘                                       │
│                                                                                          │
│  ┌────────────────────────────────────────────────────────────────────────────────────┐  │
│  │                                                                                    │  │
│  │               Namespace1                                  Namespace2               │  │
│  │ ┌────────────────────────────────────┐      ┌────────────────────────────────────┐ │  │
│  │ │                                    │      │                                    │ │  │
│  │ │   ┌────────┐        ┌─────────┐    │      │    ┌─────────┐        ┌────────┐   │ │  │
│  │ │   │ Agent1 ├────────┤ Router1 │    │      │    │ Router2 ├────────┤ Agent2 │   │ │  │
│  │ │   └────────┘        └─────────┘    │      │    └─────────┘        └────────┘   │ │  │
│  │ │ 10.0.0.2:55551   192.168.0.1:55555 │      │ 192.168.0.2:55555   10.0.0.2:55552 │ │  │
│  │ │                                    │      │                                    │ │  │
│  │ └────────────────────────────────────┘      └────────────────────────────────────┘ │  │
│  │                                                                                    │  │
│  └────────────────────────────────────────────────────────────────────────────────────┘  │
│                                                                                          │
│                                                                                          │
└──────────────────────────────────────────────────────────────────────────────────────────┘
```

The two namespaces each have their own IPs, but the real computer only has one. We need a way for the namespaces to communicate with eachother through this virtual NAT, keeping in mind that this communication may be intercepted by the real system's NAT.

There are two solutions I can think of here.
1. Create additional iptables rules on the network namespaces, something like a DNAT rule, to modify packets in some way to allow them to reach their destination (the other namespace). This could get confusing though.
2. It may be possible to get around this by creating direct veth links between the real computer and namespaces to aid with packet forwarding, but I'm not sure if this would work. Note that something like this would be required anyway for any agent inside a namespace to communicate with the testnet, since network namespaces are not automatically linked to the real network (and consequently the internet).

**Comment by CMCDragonkai** [comment]
> How should the testnet be specified as a seed node in these tests? Using `--network=testnet` isn't working for me, and specifying the deployed nodes individually using `--seed-nodes` would be too brittle since the deployed agents change relatively frequently.

We will need to solve this in MatrixAI/Polykey#403 and MatrixAI/Polykey#285.

**Comment by CMCDragonkai** [comment]
This will rebase on top of MatrixAI/Polykey#445 and will be next.

**Comment by CMCDragonkai** [comment]
Assuming all bugs are fixed.

**Comment by CMCDragonkai** [comment]
@tegefaulkes can you put up the list here in the PR description about all the tests that been disabled due to getting staging passing (from https://github.com/MatrixAI/Polykey/pull/411).

In this PR we will re-enable all of those tests, but after fixing everything.

**Comment by tegefaulkes** [comment]
Maybe they should be re-enabled in the PRs addressing the issues we disabled them for?

**Comment by tegefaulkes** [comment]
I've re-based on staging.

File: tests/integration/testnet/testnetNAT.test.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    // node signaller (and can ping back)
+  });
+  test.todo('EIM to EDM via testnet.polykey.io', async () => {
+    // Local agent behind EIM cannot ping local agent behind EDM NAT using seed
+    // node signaller (but can using relay) (and can ping back)
+  });
+  test.todo('EDM to EDM via testnet.polykey.io', async () => {
+    // Local agent behind EDM cannot ping local agent behind EDM NAT using seed
+    // node signaller (but can using relay) (and can ping back)
+  });
```
Comment (CMCDragonkai) [inline tests/integration/testnet/testnetNAT.test.ts]: These tests would all be moved to a sort of `Polykey-Simulation` repository. As we cannot control the firewall and NAT situation of the test runner in these unit tests.

**Comment by CMCDragonkai** [comment]
Remember to extract out the tests that will go into a simulation suite, and then tick off the tasks above too (and cross out ones that will be punted to the simulation suite).

**Comment by tegefaulkes** [comment]
- I changed MatrixAI/Polykey-CLI#71 to related since we want to merge this early and that's still being worked on.
- The stub tests have been moved to a new branch called `old_441_code`. 
- This PR has been reduced to 2 tests, One for a local node to connect to and ping a seed node. The other has two nodes connect to the seed node and ping each other.

**Comment by CMCDragonkai** [comment]
Don't keep it in a separate branch. We often delete branches for pruning and we may lose that code. It's better to archive that, or create a new repo just for that. Let's put it into this repo https://gitlab.com/MatrixAI/Engineering/Polykey/Polykey-Simulation.

**Comment by CMCDragonkai** [comment]
Also copy any of the remaining tasks into the README:

[ ] Verification of seed node in local keynode's NodeGraph (pk nodes getall)
[ ] Sending notification to seed node (expect failure because no trust established, but this shouldn't show up locally)
[ ] Establishing trust of seed node with local keynode (pk identities trust)
[ ] Verifying node has been added to seed node's gestalt graph after trust has been established (pk identities get/list)
[ ] Sending notification to seed node, and verifying seed node received (will need to connect to testnet and run pk notifications read)

**Comment by tegefaulkes** [comment]
I moved it to the `Polykey-Simulation` gitlab repo
